File: truevfs-kernel/truevfs-kernel-impl/src/main/scala/net/java/truevfs/kernel/impl/LockAspect.scala
/*
 * Copyright (C) 2005-2015 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.kernel.impl

import java.util.concurrent.locks._
import javax.annotation.concurrent._

/** A mixin which provides some features of its associated reentrant `lock`.
  *
  * @author Christian Schlichtherle
  */
@ThreadSafe
private trait LockAspect[+L <: Lock] {

  /** The lock with the features to provide as an aspect. */
  def lock: L

  /** Runs the given operation while the lock is held. */
  final def locked[V] = LockOn[V](lock)_
}

File: truevfs-kernel/truevfs-kernel-impl/src/main/java/net/java/truevfs/kernel/impl/LockAspect.java
/*
 * Copyright (C) 2005-2020 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.kernel.impl;

import javax.annotation.concurrent.ThreadSafe;
import java.util.concurrent.locks.Lock;

/**
 * A mixin which provides some features of its reentrant {@link #lock()}.
 *
 * @author Christian Schlichtherle
 */
@ThreadSafe
interface LockAspect<L extends Lock> {

    /**
     * Returns the lock.
     */
    L lock();

    /** Runs the given operation while holding the lock. */
    default <T, X extends Exception> T locked(Op<T, X> op) throws X { return Locks.using(lock()).call(op); }
}
